File: windows/net/samples.vb/MultiDataDisc/BurnerSettings.vb
Imports System
Imports PrimoSoftware.Burner


Namespace MultiDataDisc
	' Burn Settings
	Public Class BurnSettings
		Public SourceFolder As String
		Public VolumeLabel As String

		Public ImageType As PrimoSoftware.Burner.ImageType = ImageType.None
		Public WriteMethod As PrimoSoftware.Burner.WriteMethod = WriteMethod.DvdIncremental
		Public WriteSpeedKB As Integer = 0

		Public Simulate As Boolean = False
		Public CloseDisc As Boolean = True
		Public Eject As Boolean = True
	End Class

	Friend Class WorkerThreadContext
		Public device As Device
		Public progressForm As ProgressForm
		Public progressInfo As ProgressInfo
		Public burnerIndex As Integer
		Public burnerSettings As BurnSettings
		Public WriteRate1xKB As Double = -1

		Public Sub DataDisc_OnStatus(ByVal sender As Object, ByVal args As DataDiscStatusEventArgs)
			progressInfo.Status = GetDataDiscStatusString(args.Status)
			progressForm.UpdateProgress(progressInfo, burnerIndex)
		End Sub

		Public Sub DataDisc_OnProgress(ByVal sender As Object, ByVal args As DataDiscProgressEventArgs)
			If args.All > 0 Then
				Dim progress As Double = 100 * CDbl(args.Position) / CDbl(args.All)

				If (progress - progressInfo.Progress) > 0.1 Then
					progressInfo.ProgressStr = String.Format("{0:0.0}%", progress)
					progressInfo.Progress = progress

					If WriteRate1xKB > 0 Then
						progressInfo.WriteSpeed = String.Format("{0:0.0}x", CDbl(device.WriteTransferRate) / WriteRate1xKB)
					End If

					progressForm.UpdateProgress(progressInfo, burnerIndex)
				End If
			End If
		End Sub

		Public Sub DataDisc_OnContinueBurn(ByVal sender As Object, ByVal e As DataDiscContinueEventArgs)
			e.Continue = Not progressForm.Stopped
		End Sub

		Private Shared Function GetDataDiscStatusString(ByVal status As DataDiscStatus) As String
			Select Case status
				Case DataDiscStatus.BuildingFileSystem
					Return "Building filesystem..."
				Case DataDiscStatus.LoadingImageLayout
					Return "Loading image layout..."
				Case DataDiscStatus.WritingFileSystem
					Return "Writing filesystem..."
				Case DataDiscStatus.WritingImage
					Return "Writing image..."
				Case DataDiscStatus.CachingSmallFiles
					Return "Caching small files..."
				Case DataDiscStatus.CachingNetworkFiles
					Return "Caching network files..."
				Case DataDiscStatus.CachingCDRomFiles
					Return "Caching CDROM files..."
				Case DataDiscStatus.Initializing
					Return "Initializing and writing lead-in..."
				Case DataDiscStatus.Writing
					Return "Writing..."
				Case DataDiscStatus.WritingLeadOut
					Return "Writing lead-out and flushing cache..."
			End Select

			Return "Unknown status..."
		End Function
	End Class


	Friend Class ListItem
		Public Value As Object
		Public Description As String

		Public Sub New(ByVal nvalue As Object, ByVal description As String)
			Value = nvalue
			Me.Description = description
		End Sub

		Public Overrides Function ToString() As String
			Return Description
		End Function
	End Class


	Public Class SpeedInfo
		Public TransferRateKB As Integer
		Public TransferRate1xKB As Double
		Public Overrides Function ToString() As String
			Return String.Format("{0}x", Math.Round(CDbl(TransferRateKB) / TransferRate1xKB, 1))
		End Function
	End Class

	''' <summary>
	''' Container for device information
	''' </summary>
	Public Class DeviceInfo
		''' <summary>
		''' Device index in DeviceEnumerator
		''' </summary>
		Public Index As Integer

		Public DriveLetter As Char

		Public Title As String

		Public MediaProfile As MediaProfile

		Public MediaProfileString As String

		Public MediaFreeSpace As Long

		Public SelectedWriteSpeed As SpeedInfo

		Public MaxWriteSpeed As SpeedInfo
	End Class

	Public Class ProgressInfo
		Public DeviceTitle As String = String.Empty
		Public Status As String = String.Empty
		Public ProgressStr As String = String.Empty
		Public Progress As Double = 0.0
		Public WriteSpeed As String = String.Empty
	End Class
End Namespace

File: windows/net/samples/MultiDataDisc/BurnerSettings.cs
using System;
using PrimoSoftware.Burner;


namespace MultiDataDisc
{
	// Burn Settings
	public class BurnSettings
	{
		public string SourceFolder;
		public string VolumeLabel;

		public PrimoSoftware.Burner.ImageType ImageType = ImageType.None;
		public PrimoSoftware.Burner.WriteMethod WriteMethod = WriteMethod.DvdIncremental; 
		public int WriteSpeedKB = 0;
	
		public bool Simulate = false;
		public bool CloseDisc = true;
		public bool Eject = true;
	};

    class WorkerThreadContext
    {
        public Device device;
        public ProgressForm progressForm;
        public ProgressInfo progressInfo;
        public int burnerIndex;
        public BurnSettings burnerSettings;
        public double WriteRate1xKB = -1;

        public void DataDisc_OnStatus(Object sender, DataDiscStatusEventArgs args)
        {
            progressInfo.Status = GetDataDiscStatusString(args.Status);
            progressForm.UpdateProgress(progressInfo, burnerIndex);
        }

        public void DataDisc_OnProgress(Object sender, DataDiscProgressEventArgs args)
        {
            if (args.All > 0)
            {
                double progress = 100 * (double)args.Position / (double)args.All;

                if ((progress - progressInfo.Progress) > 0.1)
                {
                    progressInfo.ProgressStr = string.Format("{0:0.0}%", progress);
                    progressInfo.Progress = progress;

                    if (WriteRate1xKB > 0)
                    {
                        progressInfo.WriteSpeed = string.Format("{0:0.0}x", (double)device.WriteTransferRate / WriteRate1xKB);
                    }

                    progressForm.UpdateProgress(progressInfo, burnerIndex);
                }
            }
        }

        public void DataDisc_OnContinueBurn(object sender, DataDiscContinueEventArgs e)
        {
            e.Continue = !progressForm.Stopped;
        }

        static string GetDataDiscStatusString(DataDiscStatus status)
        {
            switch (status)
            {
                case DataDiscStatus.BuildingFileSystem:
                    return "Building filesystem...";
                case DataDiscStatus.LoadingImageLayout:
                    return "Loading image layout...";
                case DataDiscStatus.WritingFileSystem:
                    return "Writing filesystem...";
                case DataDiscStatus.WritingImage:
                    return "Writing image...";
                case DataDiscStatus.CachingSmallFiles:
                    return "Caching small files...";
                case DataDiscStatus.CachingNetworkFiles:
                    return "Caching network files...";
                case DataDiscStatus.CachingCDRomFiles:
                    return "Caching CDROM files...";
                case DataDiscStatus.Initializing:
                    return "Initializing and writing lead-in...";
                case DataDiscStatus.Writing:
                    return "Writing...";
                case DataDiscStatus.WritingLeadOut:
                    return "Writing lead-out and flushing cache...";
            }

            return "Unknown status...";
        }
    }


    class ListItem
    {
        public object Value;
        public string Description;

        public ListItem(object nvalue, string description)
        {
            Value = nvalue;
            Description = description;
        }

        public override string ToString()
        {
            return Description;
        }
    }


    public class SpeedInfo
    {
        public int TransferRateKB;
        public double TransferRate1xKB;
        public override string ToString()
        {
            return string.Format("{0}x", Math.Round((double)TransferRateKB / TransferRate1xKB, 1));
        }
    };

    /// <summary>
    /// Container for device information
    /// </summary>
    public class DeviceInfo
    {
        /// <summary>
        /// Device index in DeviceEnumerator
        /// </summary>
        public int Index;

        public char DriveLetter;

        public string Title;

        public MediaProfile MediaProfile;

        public string MediaProfileString;

        public long MediaFreeSpace;

        public SpeedInfo SelectedWriteSpeed;

        public SpeedInfo MaxWriteSpeed;
    };

    public class ProgressInfo
    {
        public string DeviceTitle = string.Empty;
        public string Status = string.Empty;
        public string ProgressStr = string.Empty;
        public double Progress = 0.0;
        public string WriteSpeed = string.Empty;
    }
}
